beaker/client/api_providers.py

```diff
@@ -0,0 +1,130 @@
+from enum import Enum
+from abc import ABC, abstractmethod
+from algosdk.v2client.algod import AlgodClient
+from algosdk.v2client.indexer import IndexerClient
+
+
+class Network(Enum):
+    """Provides consistent way to reference the most common network options"""
+
+    MainNet = "MainNet"
+    TestNet = "TestNet"
+    BetaNet = "BetaNet"
+    SandNet = "SandNet"
+
+
+class APIProvider(ABC):
+    """abstract class to provide interface for API providers"""
+
+    def __init__(self, network: Network):
+        self.network = network
+
+    @abstractmethod
+    def algod(self, token: str = "") -> AlgodClient:
+        """return an algod client based on the provider used and network it was initialized with"""
+        pass
+
+    @abstractmethod
+    def indexer(self, token: str = "") -> IndexerClient:
+        """return an indexer client based on the provider used and network it was initialized with"""
+        pass
+
+
+class Sandbox(APIProvider):
+    default_host = "http://localhost"
+    default_algod_port: int = 4001
+    default_indexer_port: int = 8980
+    default_token: str = "a" * 64
+
+    # purposely doesnt check which network because it may be set up to use mainnet/testnet
+
+    def algod(self, token: str = default_token) -> AlgodClient:
```

Comment:
Why are `algod` and `indexer` being redefined in the `APIProvider` subclasses? Can we not just use the methods defined in the superclass?

Reply:
sandbox doesnt care about the network used and purestake requires a diff header plus any other future providers may have their own super special way with other headers, rather than complicating the logic in the main class I figured we could just have any new api providers provide their own logic

Reply:
I'm a big believer in the DRY philosophy mainly because it can make the codebase cleaner and refactors easier. Sure the method might be a bit more complex but if there's a bugfix that's needed we update one method not "however many API providers we support" methods

Reply:
DRY is great until it isn't

A little copy/pasta here is better than a bunch of weird logic later imo

Reply:
The complexity of the logic would still be relatively low relative to the tradeoff of multiple different classes, files, and repetitions but at this point it's just a philosophical difference. If this makes more sense in the context of the Beaker codebase then it's fine with me.

Reply:
I made some changes here to implement the base case in the parent class and let the weird cases override the method optionally calling into it